src/cyclefold/circuit.rs

```diff
@@ -0,0 +1,222 @@
+//! This module defines Cyclefold stuff
+
+use bellpepper_core::{boolean::AllocatedBit, ConstraintSystem, SynthesisError};
+
+use crate::{
+  gadgets::ecc::AllocatedPoint,
+  traits::{commitment::CommitmentTrait, Engine},
+  Commitment,
+};
+
+/// TODO: docs
+pub struct CyclefoldCircuitInputs<E: Engine> {
+  commit_1: Option<Commitment<E>>,
+  commit_2: Option<Commitment<E>>,
+  result: Option<Commitment<E>>,
+  scalar: Vec<Option<bool>>,
+}
+
+impl<E: Engine> CyclefoldCircuitInputs<E> {
+  /// TODO
+  pub fn new(
+    commit_1: Option<Commitment<E>>,
+    commit_2: Option<Commitment<E>>,
+    result: Option<Commitment<E>>,
+    scalar: Vec<Option<bool>>,
+  ) -> Self {
+    Self {
+      commit_1,
+      commit_2,
+      result,
+      scalar,
+    }
+  }
+}
+
+/// TODO: docs
+pub struct CyclefoldCircuit<E: Engine> {
+  inputs: Option<CyclefoldCircuitInputs<E>>,
+}
+
+impl<E: Engine> CyclefoldCircuit<E> {
+  /// TODO: docs
+  pub fn new(inputs: Option<CyclefoldCircuitInputs<E>>) -> Self {
+    Self { inputs }
+  }
+
+  fn alloc_witness<CS: ConstraintSystem<<E as Engine>::Base>>(
+    &self,
+    mut cs: CS,
+  ) -> Result<
+    (
+      AllocatedPoint<E>,
+      AllocatedPoint<E>,
+      AllocatedPoint<E>,
+      Vec<AllocatedBit>,
+    ),
+    SynthesisError,
+  > {
+    let commit_1 = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_1"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.commit_1.map(|C_1| C_1.to_coordinates())),
+    )?;
+
+    let commit_2 = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_2"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.commit_2.map(|C_2| C_2.to_coordinates())),
+    )?;
+
+    let result = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_1 + r * C_2"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.result.map(|result| result.to_coordinates())),
+    )?;
+
+    let scalar = self
+      .inputs
+      .as_ref()
+      .ok_or(SynthesisError::AssignmentMissing)?
+      .scalar
+      .iter()
+      .enumerate()
+      .map(|(idx, b)| AllocatedBit::alloc(cs.namespace(|| format!("scalar bit {}", idx)), *b))
+      .collect::<Result<_, _>>()?;
```

Comment:
This looks like it will only allocate the bits when the inputs are present. Won't this result in non-determinism in the shape of the circuit (and hence lead to unverifiable proofs) or other runtime errors?